Physics continuous collision detection

huffman opened the pull request.

ZappoMan commented on libraries/physics/src/ObjectMotionState.cpp:
```diff
@@ -115,6 +115,29 @@ void ObjectMotionState::setMotionType(MotionType motionType) {
     _motionType = motionType;
 }
 
+void ObjectMotionState::updateCCDConfiguration() {
```
can we please give this a meaningful name and/or a comment.. CCD will not be obvious to a future developer. I know this name comes from bullet... but ObjectMotionState is our code so we can have better names.

huffman replied:
I added a comment. I think CCD is easier to read if you know what it is, `updateContinuousCollisionDetectionConfiguration` is a bit long. I can change it if that is preferred, though.

hifi-gustavo commented:
Build console output for commit: b410604475a606a53efede906cabc22f1ddacbe1 available at: https://gist.github.com/b46639c6ef3f888686c9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.dmg

hifi-gustavo commented:
Build console output for commit: c0a7da0b88b8903fa1e0aedac464b7b77ac0b35b available at: https://gist.github.com/b46639c6ef3f888686c9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.dmg

hifi-gustavo commented:
Build console output for commit: 7694958b20c31723d2cb73041f4d096ecf3739a9 available at: https://gist.github.com/b46639c6ef3f888686c9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.dmg

PhilipRosedale commented:
has this gotten tested for performance?  Where are we on merging it?

huffman commented:
@cozza13 tested it and found poor performance with the ping pong gun @ demo.

Cause:
The windmill contains a compound collision hull that is very complex, containing around 5,500 objects. When doing CCD, bullet will do collision tests between an object and every subobject of a compound object. Compound objects do maintain a dynamic AABB tree, but it isn't used when doing CCD.

Solution:
I updated bullet to do early rejection tests against the compound object's AABB tree. We should do a little more testing, but the results are good so far. I just updated this PR to include those changes to bullet. It is ready for @cozza13 to test again when it is done building.

hifi-gustavo commented:
Build console output for commit: 784df60a9750ae2ad32ec4ae9d9d2321e5024abb available at: https://gist.github.com/b46639c6ef3f888686c9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.dmg

hifi-gustavo commented:
Build console output for commit: 7e87ad5114febc0105e1155b62940271f7dff23d available at: https://gist.github.com/b46639c6ef3f888686c9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.dmg

hifi-gustavo commented:
Build console output for commit: 67d07921736a05deb5d7be780269b00ee5ef909d available at: https://gist.github.com/448258eff0d507d0e81c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.dmg

hifi-gustavo commented:
Build console output for commit: 3aa271d688c87ba4cea77190ae9505214f33a515 available at: https://gist.github.com/b46639c6ef3f888686c9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.dmg

huffman commented:
test this please

hifi-gustavo commented:
Build console output for commit: d1e9e28522c2075da9bb2e8a580cc5c3859ce7d3 available at: https://gist.github.com/b46639c6ef3f888686c9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6292.dmg

imgntn commented:
@huffman i have been standing an entity shooting really fast arrows and they all hit the entity right in front of me -- seems good to go -- is there anything else i  should test?

imgntn commented:
@huffman tested with ping pong gun in demo and don't notice any difference in performance

huffman commented:
@imgntn I want to look into one more change on the bullet side, but other than that this is mostly ready to go. Thanks for testing!

huffman commented:
I'm taking the do-not-merge off, but it still needs @AndrewMeadows to review it

AndrewMeadows commented:
This looks fine.  It uses a custom version of Bullet.  Someday (before we decide to upgrade our version of Bullet) we'll have to submit huffman's changes back into the main codebase.
